### DISCUSSION TITLE
Colors not working right after toggling `set background=light`, and also not after switching back to dark

### DISCUSSION THREAD

**Discussion by debugloop** [rebelot/heirline.nvim#172] [Q&A] [answered]
Hi folks, I use this color config together with the kanagawa theme. It's working great, but it breaks when I toggle the background setting. Whats extra weird is that it depends on modes:

1. startup, everything is nice
2. `set background=light`
3. `set background=dark`
4. normal and insert mode coloring is broken, but visual still works
5. do 2. again
6. use visual mode
7. do 3. again
8. visual mode also has broken colors

This is the relevant config, it's modeled after the cookbook example:

```lua
local colors = require("kanagawa.colors").setup()
require("heirline").load_colors(colors)
require("heirline").setup({
  tabline = {},
  statusline = {
    static = {
      mode_colors_map = {
        n = require("lualine.themes.kanagawa").normal,
        i = require("lualine.themes.kanagawa").insert,
        v = require("lualine.themes.kanagawa").visual,
        ["\22"] = require("lualine.themes.kanagawa").visual,
        c = require("lualine.themes.kanagawa").command,
        s = require("lualine.themes.kanagawa").visual,
        r = require("lualine.themes.kanagawa").replace,
        t = require("lualine.themes.kanagawa").insert,
      },
      mode_color = function(self)
        if conditions.is_active() then
          return self.mode_colors_map[self.mode:lower()]
        else
          return {
            a = "StatusLineNC",
            b = "StatusLineNC",
            c = "StatusLineNC",
          }
        end
      end,
    },
    ...
```

I'm then using the colors across all segments like this:

```lua
  hl = function(self)
    return self:mode_color().b
  end,
```

Question really is: is any of the stuff I'm doing wrong and how can I do this better? Anyone else with that issue?

**Comment by rebelot** [comment, marked as answer]
You should read the [Theming](https://github.com/rebelot/heirline.nvim/blob/master/cookbook.md#theming) section of the cookbook!

**Reply by debugloop** [reply]
Dunno how I missed it, sorry. I've actually had a similar autocommand in place for other stuff already anyways, but did not realize I should use that for heirline too.

Still, I can't quite get it to work, as now I'm stuck with the dark mode rendering of the status bar (not my winbar, that works because it only uses static hi group names). I think it's somehow in between the static section that defines the airline style color mapping, but

1. `require("kanagawa.colors").setup()` returns correctly according to `background`
2. `setup_colors` is called as outlined in the Cookbook
3. `mode_colors_map` is found and loads correctly initially...

I really don't see where I end up with an old value...

```lua
local colors = require("heirline.highlights")
local function setup_colors()
  local kanagawa_colors = require("kanagawa.colors").setup()
  kanagawa_colors.mode_colors_map = {
    n = require("lualine.themes.kanagawa").normal,
    i = require("lualine.themes.kanagawa").insert,
    v = require("lualine.themes.kanagawa").visual,
    ["\22"] = require("lualine.themes.kanagawa").visual,
    c = require("lualine.themes.kanagawa").command,
    s = require("lualine.themes.kanagawa").visual,
    r = require("lualine.themes.kanagawa").replace,
    t = require("lualine.themes.kanagawa").insert,
  }
  return kanagawa_colors
end
require("heirline").load_colors(setup_colors)
vim.api.nvim_create_autocmd("ColorScheme", {
  group = vim.api.nvim_create_augroup("Heirline", { clear = true }),
  callback = function()
    utils.on_colorscheme(setup_colors)
  end,
})
require("heirline").setup({
  tabline = {},
  statusline = {
    static = {
      mode_color = function(self)
        if conditions.is_active() then
          return colors.get_loaded_colors().mode_colors_map[self.mode:lower()]
        else
          return {
            a = "StatusLineNC",
            b = "StatusLineNC",
            c = "StatusLineNC",
          }
        end
      end,
    },
    {
    ...
```

with `hl = { fg = colors.get_loaded_colors().theme.vcs.added },` in various places as well.

Full config is [here](https://github.com/debugloop/dotfiles/blob/main/home/nvim/lua/plugins.lua#L310-L331) if needed

**Reply by debugloop** [reply]
Nevermind, solved by just using the string names like a sane person I guess :) Still curious why this did not work, but doesn't really matter 👍🏻 Sorry again for missing the section, thanks for your help 🥇
